tsdb/index/labelvalues.go

```diff
@@ -0,0 +1,332 @@
+		} else {
+			it.d.Skip(it.skip)
+		}
+
+		// Label value
+		vb := it.d.UvarintBytes()
+		v := yoloString(vb)
+
+		postingsOff := int(it.d.Uvarint64())
+		// Read from the postings table
+		postingsDec := encoding.NewDecbufAt(it.b, postingsOff, castagnoliTable)
+		_, curPostings, err := it.dec.Postings(postingsDec.Get())
+		if err != nil {
+			it.err = fmt.Errorf("decode postings: %w", err)
+			return false
+		}
+
+		it.exhausted = v == it.lastVal
+
+		isMatch := intersect(curPostings, it.postings.Clone())
+		if it.inverted {
+			isMatch = !isMatch
+		}
+		if isMatch {
+			// Make sure to allocate a new string
+			it.cur = string(vb)
+			return true
+		}
+	}
+	if it.d.Err() != nil {
+		it.err = fmt.Errorf("get postings offset entry: %w", it.d.Err())
+	}
+
+	return false
+}
+
+func (it *intersectLabelValues) At() string {
+	return it.cur
+}
+
+func (it *intersectLabelValues) Err() error {
+	return it.err
+}
+
+func (it *intersectLabelValues) Warnings() annotations.Annotations {
+	return nil
+}
+
+func (it *intersectLabelValues) Close() error {
+	return nil
+}
+
+// LabelValuesFor returns LabelValues for the given label name in the series referred to by postings.
+func (p *MemPostings) LabelValuesFor(postings Postings, name string) storage.LabelValues {
+	return p.labelValuesFor(postings, name, false)
+}
+
+// LabelValuesNotFor returns LabelValues for the given label name in the series *not* referred to by postings.
+func (p *MemPostings) LabelValuesNotFor(postings Postings, name string) storage.LabelValues {
+	return p.labelValuesFor(postings, name, true)
+}
+
+func (p *MemPostings) labelValuesFor(postings Postings, name string, inverted bool) storage.LabelValues {
+	p.mtx.RLock()
+
+	e := p.m[name]
+	if len(e) == 0 {
+		p.mtx.RUnlock()
+		return storage.EmptyLabelValues()
+	}
+
+	// With thread safety in mind and due to random key ordering in map, we have to construct the array in memory
+	vals := make([]string, 0, len(e))
+	candidates := make([]Postings, 0, len(e))
+	for val, srs := range e {
+		vals = append(vals, val)
+		candidates = append(candidates, NewListPostings(srs))
+	}
+
+	if !inverted {
+		indexes, err := FindIntersectingPostings(postings, candidates)
+		p.mtx.RUnlock()
+		if err != nil {
+			return storage.ErrLabelValues(err)
+		}
+
+		// Filter the values, keeping only those with intersecting postings
+		if len(vals) != len(indexes) {
+			slices.Sort(indexes)
+			for i, index := range indexes {
+				vals[i] = vals[index]
+			}
+			vals = vals[:len(indexes)]
+		}
+	} else {
+		// TODO: Implement findNonIntersectingPostings instead
+		indexes, err := findIntersectingPostingsMap(postings, candidates)
+		p.mtx.RUnlock()
+		if err != nil {
+			return storage.ErrLabelValues(err)
+		}
+
+		// Filter the values, keeping only those without intersecting postings
+		keep := make([]string, 0, len(vals)-len(indexes))
+		for i, val := range vals {
+			if _, ok := indexes[i]; !ok {
+				keep = append(keep, val)
+			}
+		}
+		vals = keep
```

Comment:
I think we don't need the TODO and we don't need the `findIntersectingPostingsMap` function:

```suggestion
		indexes, err := FindIntersectingPostings(postings, candidates)
		p.mtx.RUnlock()
		if err != nil {
			return storage.ErrLabelValues(err)
		}

		if len(vals) == len(indexes) {
			return storage.EmptyLabelValues()
		}

		slices.Sort(indexes)
		keep := make([]string, 0, len(vals)-len(indexes))
		idx := 0
		for i, val := range vals {
			for idx < len(indexes) && indexes[idx] < i {
				idx++
			}
			if idx == len(indexes) || indexes[idx] != i {
				keep = append(keep, val)
			}
		}
		vals = keep
```

Reply:
Thanks, nice observation.